exchanges/credentials.go

```diff
@@ -231,28 +217,27 @@ func (b *Base) SetCredentials(apiKey, apiSecret, clientID, subaccount, pemKey, o
 func (b *Base) SetAPICredentialDefaults() {
 	b.API.credMu.Lock()
 	defer b.API.credMu.Unlock()
+
 	// Exchange hardcoded settings take precedence and overwrite the config settings
 	if b.Config.API.CredentialsValidator == nil {
 		b.Config.API.CredentialsValidator = new(config.APICredentialsValidatorConfig)
 	}
-	if b.Config.API.CredentialsValidator.RequiresKey != b.API.CredentialsValidator.RequiresKey {
-		b.Config.API.CredentialsValidator.RequiresKey = b.API.CredentialsValidator.RequiresKey
-	}
 
-	if b.Config.API.CredentialsValidator.RequiresSecret != b.API.CredentialsValidator.RequiresSecret {
-		b.Config.API.CredentialsValidator.RequiresSecret = b.API.CredentialsValidator.RequiresSecret
+	// Updating credential validation requirements
+	creditRequirements := map[string]bool{
+		"RequiresKey":                b.API.CredentialsValidator.RequiresKey,
+		"RequiresSecret":             b.API.CredentialsValidator.RequiresSecret,
+		"RequiresBase64DecodeSecret": b.API.CredentialsValidator.RequiresBase64DecodeSecret,
+		"RequiresClientID":           b.API.CredentialsValidator.RequiresClientID,
+		"RequiresPEM":                b.API.CredentialsValidator.RequiresPEM,
 	}
 
-	if b.Config.API.CredentialsValidator.RequiresBase64DecodeSecret != b.API.CredentialsValidator.RequiresBase64DecodeSecret {
-		b.Config.API.CredentialsValidator.RequiresBase64DecodeSecret = b.API.CredentialsValidator.RequiresBase64DecodeSecret
-	}
-
-	if b.Config.API.CredentialsValidator.RequiresClientID != b.API.CredentialsValidator.RequiresClientID {
-		b.Config.API.CredentialsValidator.RequiresClientID = b.API.CredentialsValidator.RequiresClientID
-	}
-
-	if b.Config.API.CredentialsValidator.RequiresPEM != b.API.CredentialsValidator.RequiresPEM {
-		b.Config.API.CredentialsValidator.RequiresPEM = b.API.CredentialsValidator.RequiresPEM
+	// Updating credential validation requirements
```

Comment:
This adds more complexity I am all for that Via Negativa styles. How bout this approach?:

```diff
PS C:\Users\ohara\thrasher-corp\gocryptotrader> git diff
diff --git a/exchanges/credentials.go b/exchanges/credentials.go
index 26993d0a4..f09d9e463 100644
--- a/exchanges/credentials.go
+++ b/exchanges/credentials.go
@@ -4,7 +4,6 @@ import (
        "context"
        "errors"
        "fmt"
-       "reflect"
        "strings"

        "github.com/thrasher-corp/gocryptotrader/common/crypto"
@@ -34,7 +33,6 @@ var (
 // initializeCredentials initializes the credentials
 func (a *API) initializeCredentials() {
        a.credMu.Lock()
-       defer a.credMu.Unlock()
        if a.credentials == nil {
                a.credentials = &account.Credentials{}
        }
@@ -44,30 +42,35 @@ func (a *API) initializeCredentials() {
 func (a *API) SetKey(key string) {
        a.initializeCredentials()
        a.credentials.Key = key
+       a.credMu.Unlock()
 }

 // SetSecret sets new secret for the default credentials
 func (a *API) SetSecret(secret string) {
        a.initializeCredentials()
        a.credentials.Secret = secret
+       a.credMu.Unlock()
 }

 // SetClientID sets new clientID for the default credentials
 func (a *API) SetClientID(clientID string) {
        a.initializeCredentials()
        a.credentials.ClientID = clientID
+       a.credMu.Unlock()
 }

 // SetPEMKey sets pem key for the default credentials
 func (a *API) SetPEMKey(pem string) {
        a.initializeCredentials()
        a.credentials.PEMKey = pem
+       a.credMu.Unlock()
 }

 // SetSubAccount sets sub account for the default credentials
 func (a *API) SetSubAccount(sub string) {
        a.initializeCredentials()
        a.credentials.SubAccount = sub
+       a.credMu.Unlock()
 }

 // CheckCredentials checks to see if the required fields have been set before
@@ -190,6 +193,7 @@ func (b *Base) VerifyAPICredentials(creds *account.Credentials) error {
 // SetCredentials is a method that sets the current API keys for the exchange
 func (b *Base) SetCredentials(apiKey, apiSecret, clientID, subaccount, pemKey, oneTimePassword string) {
        b.API.initializeCredentials()
+       defer b.API.credMu.Unlock()
        b.API.credentials.Key = apiKey
        b.API.credentials.ClientID = clientID
        b.API.credentials.SubAccount = subaccount
@@ -223,22 +227,7 @@ func (b *Base) SetAPICredentialDefaults() {
                b.Config.API.CredentialsValidator = new(config.APICredentialsValidatorConfig)
        }

-       // Updating credential validation requirements
-       creditRequirements := map[string]bool{
-               "RequiresKey":                b.API.CredentialsValidator.RequiresKey,
-               "RequiresSecret":             b.API.CredentialsValidator.RequiresSecret,
-               "RequiresBase64DecodeSecret": b.API.CredentialsValidator.RequiresBase64DecodeSecret,
-               "RequiresClientID":           b.API.CredentialsValidator.RequiresClientID,
-               "RequiresPEM":                b.API.CredentialsValidator.RequiresPEM,
-       }
-
-       // Updating credential validation requirements
-       for requirement, value := range creditRequirements {
-               field := reflect.ValueOf(b.Config.API.CredentialsValidator).Elem().FieldByName(requirement)
-               if field.IsValid() && field.CanSet() && field.Bool() != value {
-                       field.SetBool(value)
-               }
-       }
+       *b.Config.API.CredentialsValidator = b.API.CredentialsValidator
 }

 // IsWebsocketAuthenticationSupported returns whether the exchange supports
diff --git a/exchanges/credentials_test.go b/exchanges/credentials_test.go
index b7de75718..6b5a97bbb 100644
--- a/exchanges/credentials_test.go
+++ b/exchanges/credentials_test.go
@@ -186,7 +186,7 @@ func TestVerifyAPICredentials(t *testing.T) {
        setupBase := func(tData *tester) *Base {
                b := &Base{
                        API: API{
-                               CredentialsValidator: CredentialsValidator{
+                               CredentialsValidator: config.APICredentialsValidatorConfig{
                                        RequiresKey:                tData.RequiresKey,
                                        RequiresSecret:             tData.RequiresSecret,
                                        RequiresClientID:           tData.RequiresClientID,
@@ -244,7 +244,7 @@ func TestCheckCredentials(t *testing.T) {
                        name: "Test credentials failure",
                        base: &Base{
                                API: API{
-                                       CredentialsValidator: CredentialsValidator{RequiresKey: true},
+                                       CredentialsValidator: config.APICredentialsValidatorConfig{RequiresKey: true},
                                        credentials:          &account.Credentials{OneTimePassword: "wow"},
                                },
                        },
@@ -255,7 +255,7 @@ func TestCheckCredentials(t *testing.T) {
                        base: &Base{
                                LoadedByConfig: true,
                                API: API{
-                                       CredentialsValidator: CredentialsValidator{RequiresKey: true},
+                                       CredentialsValidator: config.APICredentialsValidatorConfig{RequiresKey: true},
                                        credentials:          &account.Credentials{Key: "k3y"},
                                },
                        },
@@ -267,7 +267,7 @@ func TestCheckCredentials(t *testing.T) {
                                LoadedByConfig: true,
                                API: API{
                                        AuthenticatedSupport: true,
-                                       CredentialsValidator: CredentialsValidator{RequiresKey: true},
+                                       CredentialsValidator: config.APICredentialsValidatorConfig{RequiresKey: true},
                                        credentials:          &account.Credentials{},
                                },
                        },
@@ -277,7 +277,7 @@ func TestCheckCredentials(t *testing.T) {
                        name: "Test base64 decoded invalid credentials",
                        base: &Base{
                                API: API{
-                                       CredentialsValidator: CredentialsValidator{RequiresBase64DecodeSecret: true},
+                                       CredentialsValidator: config.APICredentialsValidatorConfig{RequiresBase64DecodeSecret: true},
                                        credentials:          &account.Credentials{Secret: "invalid"},
                                },
                        },
@@ -287,7 +287,7 @@ func TestCheckCredentials(t *testing.T) {
                        name: "Test base64 decoded valid credentials",
                        base: &Base{
                                API: API{
-                                       CredentialsValidator: CredentialsValidator{RequiresBase64DecodeSecret: true},
+                                       CredentialsValidator: config.APICredentialsValidatorConfig{RequiresBase64DecodeSecret: true},
                                        credentials:          &account.Credentials{Secret: "aGVsbG8gd29ybGQ="},
                                },
                        },
@@ -299,7 +299,7 @@ func TestCheckCredentials(t *testing.T) {
                        base: &Base{
                                API: API{
                                        AuthenticatedSupport: true,
-                                       CredentialsValidator: CredentialsValidator{RequiresKey: true},
+                                       CredentialsValidator: config.APICredentialsValidatorConfig{RequiresKey: true},
                                        credentials:          &account.Credentials{Key: "k3y"},
                                },
                        },
diff --git a/exchanges/exchange_types.go b/exchanges/exchange_types.go
index 25ac66b6b..982dfe4b5 100644
--- a/exchanges/exchange_types.go
+++ b/exchanges/exchange_types.go
@@ -205,17 +205,7 @@ type API struct {
        credentials *account.Credentials
        credMu      sync.RWMutex

-       CredentialsValidator CredentialsValidator
-}
-
-// CredentialsValidator determines what is required
-// to make authenticated requests for an exchange
-type CredentialsValidator struct {
-       RequiresPEM                bool
-       RequiresKey                bool
-       RequiresSecret             bool
-       RequiresClientID           bool
-       RequiresBase64DecodeSecret bool
+       CredentialsValidator config.APICredentialsValidatorConfig
 }

 // Base stores the individual exchange information
```

So we actually remove the type declaration for `CredentialsValidator` and use the config type, then we can just assign values cutting out a bunch of code by doing `*b.Config.API.CredentialsValidator = b.API.CredentialsValidator`

Reply:
Looks great

Reply:
I don't think we should have the locking and unlocking of a mutex handled in separate functions. I'd be more supportive of handling like

```go
// initializeCredentials initializes the credentials, requires a.credMu lock before calling
func (a *API) initializeCredentials() {
	if a.credentials == nil {
		a.credentials = &account.Credentials{}
	}
}

// SetKey sets new key for the default credentials
func (a *API) SetKey(key string) {
	a.credMu.Lock()
	defer a.credMu.Unlock()
	a.initializeCredentials()
	a.credentials.Key = key
}

...
```

I also don't really see a need for `credentials` to be stored in `API` as a pointer given every exchange we support has API credentials. It would remove the need for `initializeCredentials()` but that can be considered at some other point